File: apps/desktop/src/services/client/resources/label.ts
import type { Label } from "../../../types";
import { call } from "../invoke";
import type { CreateLabelBody } from "../types";

export class LabelResource {
  list(ws: string, type = "task") {
    return call<Label[]>("label_list", { workspaceId: ws, labelType: type });
  }

  create(ws: string, body: CreateLabelBody) {
    return call<{ id: string }>("label_create", {
      workspaceId: ws,
      name: body.name,
      labelType: body.type ?? "task",
    });
  }

  attach(ws: string, taskId: string, labelId: string) {
    return call<void>("label_attach", { workspaceId: ws, taskId, labelId });
  }

  detach(ws: string, taskId: string, labelId: string) {
    return call<void>("label_detach", { workspaceId: ws, taskId, labelId });
  }
}

File: apps/desktop/src-tauri/src/grpc/label.rs
use modula_client::ModulaClient;
use modula_types::Label;
use serde_json::{json, Value};
use tauri::State;

#[tauri::command]
pub async fn label_list(
    engine: State<'_, ModulaClient>,
    workspace_id: String,
    label_type: String,
) -> Result<Vec<Label>, String> {
    Ok(engine.list_labels(&workspace_id, &label_type).await?)
}

#[tauri::command]
pub async fn label_create(
    engine: State<'_, ModulaClient>,
    workspace_id: String,
    name: String,
    label_type: String,
) -> Result<Value, String> {
    let id = engine
        .create_label(&workspace_id, &name, &label_type)
        .await?;
    Ok(json!({ "id": id }))
}

#[tauri::command]
pub async fn label_attach(
    engine: State<'_, ModulaClient>,
    workspace_id: String,
    task_id: String,
    label_id: String,
) -> Result<(), String> {
    engine
        .attach_label(&workspace_id, &task_id, &label_id)
        .await?;
    Ok(())
}

#[tauri::command]
pub async fn label_detach(
    engine: State<'_, ModulaClient>,
    workspace_id: String,
    task_id: String,
    label_id: String,
) -> Result<(), String> {
    engine
        .detach_label(&workspace_id, &task_id, &label_id)
        .await?;
    Ok(())
}
